src/serializer/ResidualHeapGraphGenerator.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import type { Logger } from "./logger.js";
+import type { Modules } from "./modules.js";
+import type { Realm } from "../realm.js";
+import type { Effects } from "../realm.js";
+import type { ObjectRefCount } from "./types.js";
+
+import invariant from "../invariant.js";
+import {
+  Value,
+  EmptyValue,
+  FunctionValue,
+  AbstractValue,
+  SymbolValue,
+  ProxyValue,
+  ObjectValue,
+} from "../values/index.js";
+import { ResidualHeapInspector } from "./ResidualHeapInspector.js";
+import { ResidualHeapVisitor } from "./ResidualHeapVisitor.js";
+
+type Edge = {
+  fromId: number,
+  toId: number,
+};
+
+/**
+ * Generate a visualizable objects graph for Prepack heap.
+ */
+export class ResidualHeapGraphGenerator extends ResidualHeapVisitor {
+  constructor(
+    realm: Realm,
+    logger: Logger,
+    modules: Modules,
+    additionalFunctionValuesAndEffects: Map<FunctionValue, Effects>,
+    valueToEdgeRecord: Map<Value, ObjectRefCount>
+  ) {
+    super(realm, logger, modules, additionalFunctionValuesAndEffects);
+    this._valueToEdgeRecord = valueToEdgeRecord;
+    this._visitedValues = new Set();
+    this._valueIds = new Map();
+    this._idSeed = 0;
+    this._ancestors = [];
+    this._edges = [];
+  }
+
+  _valueToEdgeRecord: Map<Value, ObjectRefCount>;
+  _valueIds: Map<Value, number>;
+  _idSeed: number;
+  _visitedValues: Set<Value>;
+  _ancestors: Array<Value>;
+  _edges: Array<Edge>;
+
+  // Override.
+  preProcessValue(val: Value): boolean {
+    if (this._shouldIgnore(val)) {
+      return true;
+    }
+    this._updateEdge(val);
+
+    if (this._visitedValues.has(val)) {
+      return false; // Already visited.
+    }
+    this._visitedValues.add(val);
+    return true;
+  }
+
+  // Override.
+  postProcessValue(val: Value): void {
+    if (this._shouldIgnore(val)) {
+      return;
+    }
+    invariant(this._ancestors.length > 0);
+    this._ancestors.pop();
+  }
+
+  _getValueId(val: Value): number {
+    let id = this._valueIds.get(val);
+    if (!id) {
+      this._valueIds.set(val, ++this._idSeed);
+      id = this._idSeed;
+    }
+    return id;
+  }
+
+  _shouldIgnore(val: Value): boolean {
+    return val instanceof EmptyValue || val.isIntrinsic() || ResidualHeapInspector.isLeaf(val);
+  }
+
+  _updateEdge(val: Value) {
+    if (this._ancestors.length > 0) {
+      const parent = this._ancestors[this._ancestors.length - 1];
```

Comment:
Perhaps this should be named "predecessor"?

Reply:
I am modeling the object reference relationship as parent=>child here. Is that not correct mental model for it? I would like to avoid "predecessor" because it may be confused with the term "predecessor" during graph's traversal which may not be the parent node if BFS is used.